python/eos/analysis_file.py

```diff
@@ -155,44 +155,134 @@ def _sanitize_params(command, params):
+                    if type(value) is str:
+                        arguments.update({key: value.format(**_arguments)})
+                    else:
+                        arguments.update({key: value})
+
+                result.append(arguments)
+
+            return result
+
+        def _handle_single_step(step_desc, step_arguments):
+            step_name = step_desc['name']
+            tasks     = step_desc['tasks']
+
+            for task_desc in tasks:
+                task = task_desc['task']
+                if task not in _tasks.keys():
+                    raise ValueError(f'Unknown task \"{task}\" encountered in step {step_name}')
+
+                arguments = { 'analysis_file': self }
+                for key, value in step_arguments.items():
+                    if key in _tasks.keys():
+                        continue
+
+                    arguments[key] = value
+
+                if 'arguments' in task_desc:
+                    arguments.update(task_desc['arguments'])
+
+                if task in step_arguments:
+                    arguments.update(step_arguments[task])
+
+                if base_directory:
+                    arguments.update({ 'base_directory': base_directory })
+
+                return [(step_name, step_desc, task, a) for a in _expand_arguments(step_name, task, arguments)]
+
+        # main part starts here
+
+        # check inputs
+        for step_idx, step_desc in enumerate(self._steps):
+            if type(step_desc) is not dict:
+                raise ValueError(f'Description of step #{step_idx} is not a key/value map')
+
+            if 'name' not in step_desc:
+                raise ValueError(f'Description of step #{step_idx} requires a name')
+
+            if 'tasks' not in step_desc:
+                raise ValueError('Description of step #{step_idx} ({step["name"]}) requires a list of tasks')
+
+            if 'iterations' in step_desc:
+                if type(step_desc['iterations']) is not list:
+                    raise ValueError(f'Description of step #{step_idx} ({step["name"]} expects a list of iterations')
+
+        # resolve dependencies
+        graph = nx.MultiDiGraph()
+        graph.add_nodes_from([step['name'] for step in self._steps])
+        for step in self._steps:
+            if not 'depends-on' in step:
+                continue
+
+            for dep in step['depends-on']:
+                graph.add_edge(dep, step['name'])
+        steps = list(nx.topological_sort(graph))
+
+        # return steps and theirs tasks in the order imposed by the dependency graph
+        result = []
+        for step in steps:
+            step_desc = { sd['name']: sd for sd in self._steps }[step]
+            iterations = [{}]
+            if 'iterations' in step_desc:
+                iterations = step_desc['iterations']
+
+            for iteration_arguments in iterations:
+                result.extend(_handle_single_step(step_desc, iteration_arguments))
```

Comment:
Where is this used? I couldn't find, I was expecting this to appear in the `Executor`, no?